README.rst
Foris controller netmetr module
===============================
This is a netmetr module for foris-controller

Requirements
============

* foris-controller
* netmetr command

Installation
============

	``python setup.py install``


CHANGELOG.rst
0.4.2 (2018-11-30)
------------------

* setup.py: cleanup + PEP508 updated

0.4.1 (2018-08-17)
------------------

* python3 division fix

0.4 (2018-08-13)
----------------

* reflect api changes

0.3 (2018-05-24)
----------------

* more resilent data parsing

0.2 (2017-12-13)
----------------

* reflect foris-controller updates

0.1.1 (2017-12-04)
------------------

* default sync_code value fix

0.1 (2017-11-16)
----------------

* initial version


tests/conftest.py
#
# foris-controller-netmetr-module
# Copyright (C) 2017 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

import pytest
import os

# load common fixtures
from foris_controller_testtools.fixtures import (
    ubusd_acl_path, uci_config_default_path, cmdline_script_root,
    controller_modules, extra_module_paths, message_bus, backend
)


@pytest.fixture(scope="session")
def ubusd_acl_path():
    return os.path.join(os.path.dirname(os.path.realpath(__file__)), "ubus-acl")


@pytest.fixture(scope="session")
def uci_config_default_path():
    return os.path.join(
        os.path.dirname(os.path.realpath(__file__)), "uci_configs"
    )


@pytest.fixture(scope="session")
def cmdline_script_root():
    return os.path.join(
        os.path.dirname(os.path.realpath(__file__)), "test_root"
    )


@pytest.fixture(scope="module")
def controller_modules():
    return ["netmetr"]


def pytest_addoption(parser):
    parser.addoption(
        "--backend", action="append",
        default=[],
        help=("Set test backend here. available values = (mock, openwrt)")
    )
    parser.addoption(
        "--debug-output", action="store_true",
        default=False,
        help=("Whether show output of foris-controller cmd")
    )


def pytest_generate_tests(metafunc):
    if 'backend' in metafunc.fixturenames:
        backend = set(metafunc.config.option.backend)
        if not backend:
            backend = ['mock']
        metafunc.parametrize("backend_param", backend, scope='module')


tests/fixtures.py
#
# foris-controller-netmetr-module
# Copyright (C) 2017 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#


import json
import os
import pytest
import socket
import subprocess
import time
import struct
import sys
import uuid

from multiprocessing import Process, Value, Lock

if sys.version_info < (3, 0):
    import SocketServer
else:
    import socketserver
    SocketServer = socketserver

SOCK_PATH = "/tmp/foris-controller-test.soc"
UBUS_PATH = "/tmp/ubus-foris-controller-test.soc"
NOTIFICATION_SOCK_PATH = "/tmp/foris-controller-notifications-test.soc"
NOTIFICATIONS_OUTPUT_PATH = "/tmp/foris-controller-notifications-test.json"

notifications_lock = Lock()


def chunks(data, size):
    for i in range(0, len(data), size):
        yield data[i:i + size]


@pytest.fixture(scope="session")
def ubusd_test():
    ubusd_instance = subprocess.Popen(["ubusd", "-A", "tests/ubus-acl", "-s", UBUS_PATH])
    yield ubusd_instance
    ubusd_instance.kill()
    try:
        os.unlink(SOCK_PATH)
    except:
        pass


def ubus_notification_listener(exiting):
    import prctl
    import signal
    prctl.set_pdeathsig(signal.SIGKILL)
    import ubus
    ubus.connect(UBUS_PATH)
    global notifications_lock

    try:
        os.unlink(NOTIFICATIONS_OUTPUT_PATH)
    except OSError:
        if os.path.exists(NOTIFICATIONS_OUTPUT_PATH):
            raise

    with open(NOTIFICATIONS_OUTPUT_PATH, "w") as f:
        f.flush()

        def handler(module, data):
            module_name = module[len("foris-controller-"):]
            with notifications_lock:
                f.write(json.dumps({
                    "module": module_name,
                    "kind": "notification",
                    "action": data["action"],
                    "data": data["data"],
                }) + "\n")
                f.flush()

        ubus.listen(("foris-controller-*", handler))
        while True:
            ubus.loop(200)
            if exiting.value:
                break


def unix_notification_listener():
    import prctl
    import signal
    from threading import Lock
    lock = Lock()
    prctl.set_pdeathsig(signal.SIGKILL)
    global notifications_lock

    try:
        os.unlink(NOTIFICATION_SOCK_PATH)
    except OSError:
        if os.path.exists(NOTIFICATION_SOCK_PATH):
            raise
    try:
        os.unlink(NOTIFICATIONS_OUTPUT_PATH)
    except OSError:
        if os.path.exists(NOTIFICATIONS_OUTPUT_PATH):
            raise

    class Server(SocketServer.ThreadingMixIn, SocketServer.UnixStreamServer):
        pass

    with open(NOTIFICATIONS_OUTPUT_PATH, "w") as f:
        f.flush()

        class Handler(SocketServer.StreamRequestHandler):
            def handle(self):
                while True:
                    length_raw = self.rfile.read(4)
                    if len(length_raw) != 4:
                        break
                    length = struct.unpack("I", length_raw)[0]
                    data = self.rfile.read(length)
                    with lock:
                        with notifications_lock:
                            f.write(data + "\n")
                            f.flush()

        server = Server(NOTIFICATION_SOCK_PATH, Handler)
        server.serve_forever()


class Infrastructure(object):
    def __init__(self, name, backend_name, debug_output=False):
        try:
            os.unlink(SOCK_PATH)
        except:
            pass

        self.name = name
        self.backend_name = backend_name
        if name not in ["unix-socket", "ubus"]:
            raise NotImplementedError()
        if backend_name not in ["openwrt", "mock"]:
            raise NotImplementedError()

        self.sock_path = SOCK_PATH
        if name == "ubus":
            self.sock_path = UBUS_PATH
            while not os.path.exists(self.sock_path):
                time.sleep(0.3)

        kwargs = {}
        if not debug_output:
            devnull = open(os.devnull, 'wb')
            kwargs['stderr'] = devnull
            kwargs['stdout'] = devnull

        self._exiting = Value('i', 0)
        self._exiting.value = False

        if name == "unix-socket":
            self.listener = Process(target=unix_notification_listener, args=tuple())
            self.listener.start()
        elif name == "ubus":
            self.listener = Process(target=ubus_notification_listener, args=(self._exiting, ))
            self.listener.start()

        args = [
            "foris-controller",
            "-m", ",".join(["netmetr"]),
            "-d", "-b", backend_name, name, "--path", self.sock_path
        ]

        if name == "unix-socket":
            args.append("--notifications-path")
            args.append(NOTIFICATION_SOCK_PATH)
            self.notification_sock_path = NOTIFICATION_SOCK_PATH
        else:
            self.notification_sock_path = self.sock_path

        self.server = subprocess.Popen(args, **kwargs)

    def exit(self):
        self._exiting.value = True
        self.server.kill()
        self.listener.terminate()
        try:
            os.unlink(NOTIFICATIONS_OUTPUT_PATH)
        except OSError:
            pass

    def process_message(self, data):
        if self.name == "unix-socket":
            while not os.path.exists(self.sock_path):
                time.sleep(1)
            sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
            sock.connect(self.sock_path)
            data = json.dumps(data).encode("utf8")
            length_bytes = struct.pack("I", len(data))
            sock.sendall(length_bytes + data)

            length = struct.unpack("I", sock.recv(4))[0]
            received = sock.recv(length)
            recv_len = len(received)
            while recv_len < length:
                received += sock.recv(length)
                recv_len = len(received)

            return json.loads(received.decode("utf8"))

        elif self.name == "ubus":
            import ubus
            module = "foris-controller-%s" % data.get("module", "?")
            wait_process = subprocess.Popen(
                ["ubus", "wait_for", module, "-s", self.sock_path])
            wait_process.wait()
            if not ubus.get_connected():
                ubus.connect(self.sock_path)
            function = data.get("action", "?")
            inner_data = data.get("data", {})
            dumped_data = json.dumps(inner_data)
            request_id = str(uuid.uuid4())
            if len(dumped_data) > 512 * 1024:
                for data_part in chunks(dumped_data, 512 * 1024):
                    ubus.call(module, function, {
                        "data": {}, "final": False, "multipart": True,
                        "request_id": request_id, "multipart_data": data_part,
                    })

                res = ubus.call(module, function, {
                    "data": {}, "final": True, "multipart": True,
                    "request_id": request_id, "multipart_data": "",
                })

            else:
                res = ubus.call(module, function, {
                    "data": inner_data, "final": True, "multipart": False,
                    "request_id": request_id, "multipart_data": "",
                })

            ubus.disconnect()
            return {
                u"module": data["module"],
                u"action": data["action"],
                u"kind": u"reply",
                u"data": json.loads("".join([e["data"] for e in res])),
            }

        raise NotImplementedError()

    def get_notifications(self, old_data=None):
        while not os.path.exists(NOTIFICATIONS_OUTPUT_PATH):
            time.sleep(0.2)

        global notifications_lock

        while True:
            with notifications_lock:
                with open(NOTIFICATIONS_OUTPUT_PATH) as f:
                    data = f.readlines()
                    last_data = [json.loads(e.strip()) for e in data]
                    if not old_data == last_data:
                        break
        return last_data


@pytest.fixture(scope="module")
def backend(backend_param):
    return backend_param


@pytest.fixture(params=["unix-socket", "ubus"], scope="module")
def infrastructure(request, backend):
    instance = Infrastructure(
        request.param, backend, request.config.getoption("--debug-output"))
    yield instance
    instance.exit()


@pytest.fixture(scope="module")
def infrastructure_unix_socket(request, backend):
    instance = Infrastructure(
        "unix-socket", backend, request.config.getoption("--debug-output"))
    yield instance
    instance.exit()


tests/test_netmetr.py
#
# foris-controller-netmetr-module
# Copyright (C) 2017 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

import pytest
import textwrap

from foris_controller_testtools.fixtures import (
    backend, infrastructure, ubusd_test, uci_configs_init, FILE_ROOT_PATH
)

from foris_controller_testtools.utils import FileFaker


@pytest.fixture(scope="function")
def netmetr_data(request):
    content = """\
        {
          "error": [],
          "history": [
            {
              "speed_download": "920",
              "speed_upload": "910",
              "ping_shortest_classification": 3,
              "test_type_id": 1,
              "speed_download_classification": 3,
              "ping": "0.78",
              "ping_shortest": "0.78",
              "time_string": "Aug 17, 2018 2:17:00 AM",
              "speed_upload_classification": 3,
              "time": 1534472220000,
              "ping_classification": 3,
              "timezone": "CEST",
              "model": "Turris",
              "test_uuid": "28a371fc-e310-48b5-9804-fe12cbfb692a",
              "network_type": "LAN"
            },
            {
              "speed_download": "880",
              "speed_upload": "890",
              "ping_shortest_classification": 3,
              "test_type_id": 1,
              "speed_download_classification": 3,
              "ping": "0.76",
              "ping_shortest": "0.76",
              "time_string": "Aug 16, 2018 2:28:31 AM",
              "speed_upload_classification": 3,
              "time": 1534386511000,
              "ping_classification": 3,
              "timezone": "CEST",
              "model": "Turris",
              "test_uuid": "a2148440-89a6-4ef3-a184-8f2009fc2939",
              "network_type": "LAN"
            },
            {
              "speed_download": "910",
              "speed_upload": "910",
              "ping_shortest_classification": 3,
              "test_type_id": 1,
              "speed_download_classification": 3,
              "ping": "0.71",
              "ping_shortest": "0.71",
              "time_string": "Aug 15, 2018 2:27:46 AM",
              "speed_upload_classification": 3,
              "time": 1534300066000,
              "ping_classification": 3,
              "timezone": "CEST",
              "model": "Turris",
              "test_uuid": "7f004e22-cd8a-470b-a1b0-9139c68632c9",
              "network_type": "LAN"
            },
            {
              "speed_download": "930",
              "speed_upload": "920",
              "ping_shortest_classification": 3,
              "test_type_id": 1,
              "speed_download_classification": 3,
              "ping": "0.77",
              "ping_shortest": "0.77",
              "time_string": "Aug 14, 2018 2:26:43 AM",
              "speed_upload_classification": 3,
              "time": 1534213603000,
              "ping_classification": 3,
              "timezone": "CEST",
              "model": "Turris",
              "test_uuid": "19a7ecc9-034d-40da-93a3-f002e8843057",
              "network_type": "LAN"
            },
            {
              "speed_download": "920",
              "speed_upload": "920",
              "ping_shortest_classification": 3,
              "test_type_id": 1,
              "speed_download_classification": 3,
              "ping": "0.71",
              "ping_shortest": "0.71",
              "time_string": "Aug 13, 2018 1:45:24 PM",
              "speed_upload_classification": 3,
              "time": 1534167924000,
              "ping_classification": 3,
              "timezone": "CEST",
              "model": "Turris",
              "test_uuid": "66ef8238-2bf6-4a2c-8a22-99b14c01b6cb",
              "network_type": "LAN"
            },
            {
              "speed_download": "920",
              "speed_upload": "930",
              "ping_shortest_classification": 3,
              "test_type_id": 1,
              "speed_download_classification": 3,
              "ping": "0.79",
              "ping_shortest": "0.79",
              "time_string": "Aug 13, 2018 4:50:01 AM",
              "speed_upload_classification": 3,
              "time": 1534135801000,
              "ping_classification": 3,
              "timezone": "CEST",
              "model": "Turris",
              "test_uuid": "a498c3a9-c077-4d3b-bb42-aa0f760558df",
              "network_type": "LAN"
            },
            {
              "speed_download": "910",
              "speed_upload": "910",
              "ping_shortest_classification": 3,
              "test_type_id": 1,
              "speed_download_classification": 3,
              "ping": "0.74",
              "ping_shortest": "0.74",
              "time_string": "Aug 13, 2018 2:24:07 AM",
              "speed_upload_classification": 3,
              "time": 1534127047000,
              "ping_classification": 3,
              "timezone": "CEST",
              "model": "Turris",
              "test_uuid": "f93d5d91-dc47-4b64-a957-26cb24f3251a",
              "network_type": "LAN"
            },
            {
              "speed_download": "880",
              "speed_upload": "900",
              "ping_shortest_classification": 3,
              "test_type_id": 1,
              "speed_download_classification": 3,
              "ping": "0.74",
              "ping_shortest": "0.74",
              "time_string": "Aug 12, 2018 2:18:12 AM",
              "speed_upload_classification": 3,
              "time": 1534040292000,
              "ping_classification": 3,
              "timezone": "CEST",
              "model": "Turris",
              "test_uuid": "c7589adb-1f44-4710-8511-4d0cc6c96a6b",
              "network_type": "LAN"
            },
            {
              "speed_download": "800",
              "speed_upload": "770",
              "ping_shortest_classification": 3,
              "test_type_id": 1,
              "speed_download_classification": 3,
              "ping": "0.73",
              "ping_shortest": "0.73",
              "time_string": "Aug 11, 2018 2:25:38 AM",
              "speed_upload_classification": 3,
              "time": 1533954338000,
              "ping_classification": 3,
              "timezone": "CEST",
              "model": "Turris",
              "test_uuid": "73f2469d-0e09-4d56-8536-7bfe9af9f9b6",
              "network_type": "LAN"
            },
            {
              "speed_download": "920",
              "speed_upload": "890",
              "ping_shortest_classification": 3,
              "test_type_id": 1,
              "speed_download_classification": 3,
              "ping": "0.64",
              "ping_shortest": "0.64",
              "time_string": "Aug 10, 2018 2:11:35 AM",
              "speed_upload_classification": 3,
              "time": 1533867095000,
              "ping_classification": 3,
              "timezone": "CEST",
              "model": "Turris",
              "test_uuid": "586bd913-abd0-433b-bef4-745b3c4ad034",
              "network_type": "LAN"
            }
          ]
        }"""
    with FileFaker(
        FILE_ROOT_PATH, "/tmp/netmetr-history.json",
        False, textwrap.dedent(content)
    ) as f:
        yield f, content


def test_settings(uci_configs_init, infrastructure, ubusd_test):
    notifications = infrastructure.get_notifications()

    res = infrastructure.process_message({
        "module": "netmetr",
        "action": "get_settings",
        "kind": "request",
    })
    assert set(res["data"].keys()) == {"autostart_enabled", "hours_to_run", "sync_code"}
    sync_code = res["data"]["sync_code"]
    new_autostart_enabled = not res["data"]["autostart_enabled"]
    new_hours_to_run = [e for e in range(24) if e not in res["data"]["hours_to_run"]]

    res = infrastructure.process_message({
        "module": "netmetr",
        "action": "update_settings",
        "kind": "request",
        "data": {
            "autostart_enabled": new_autostart_enabled,
            "hours_to_run": new_hours_to_run,
        }
    })
    assert res["data"] == {"result": True}

    notifications = infrastructure.get_notifications(notifications)
    assert notifications[-1] == {
        u"module": u"netmetr",
        u"action": u"update_settings",
        u"kind": u"notification",
        u"data": {u"autostart_enabled": new_autostart_enabled, u"hours_to_run": new_hours_to_run},
    }

    res = infrastructure.process_message({
        "module": "netmetr",
        "action": "get_settings",
        "kind": "request",
    })
    assert res == {
        u"module": u"netmetr",
        u"action": u"get_settings",
        u"kind": u"reply",
        u"data": {
            u"sync_code": sync_code,
            u"autostart_enabled": new_autostart_enabled,
            u"hours_to_run": new_hours_to_run
        },
    }

    # command without hours_to_run
    new_autostart_enabled = not new_autostart_enabled
    res = infrastructure.process_message({
        "module": "netmetr",
        "action": "update_settings",
        "kind": "request",
        "data": {
            "autostart_enabled": new_autostart_enabled,
        }
    })
    assert res["data"] == {"result": True}

    notifications = infrastructure.get_notifications(notifications)
    assert notifications[-1] == {
        u"module": u"netmetr",
        u"action": u"update_settings",
        u"kind": u"notification",
        u"data": {u"autostart_enabled": new_autostart_enabled},
    }

    res = infrastructure.process_message({
        "module": "netmetr",
        "action": "get_settings",
        "kind": "request",
    })
    assert res == {
        u"module": u"netmetr",
        u"action": u"get_settings",
        u"kind": u"reply",
        u"data": {
            u"sync_code": sync_code,
            u"autostart_enabled": new_autostart_enabled,
            u"hours_to_run": new_hours_to_run
        },
    }


def test_get_data(uci_configs_init, infrastructure, ubusd_test, netmetr_data):
    res = infrastructure.process_message({
        "module": "netmetr",
        "action": "get_data",
        "kind": "request",
    })
    assert set(res["data"].keys()) == {"status", "performed_tests"}


def test_download_data(uci_configs_init, infrastructure, ubusd_test):
    res = infrastructure.process_message({
        "module": "netmetr",
        "action": "download_data",
        "kind": "request",
    })
    assert set(res["data"].keys()) == {"async_id"}


def test_measure_and_download_data(uci_configs_init, infrastructure, ubusd_test):
    res = infrastructure.process_message({
        "module": "netmetr",
        "action": "measure_and_download_data",
        "kind": "request",
    })
    assert set(res["data"].keys()) == {"async_id"}


foris_controller_netmetr_module/__init__.py
#
# foris-controller-netmetr-module
# Copyright (C) 2017 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

__version__ = "0.4.2"


foris_controller_modules/netmetr/handlers/openwrt.py
#
# foris-controller-netmetr-module
# Copyright (C) 2017 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

import logging

from foris_controller.handler_base import BaseOpenwrtHandler
from foris_controller.utils import logger_wrapper

from foris_controller_backends.netmetr import NetmetrUci, NetmetrDataFile, NetmetrCmds

from .. import Handler

logger = logging.getLogger(__name__)


class OpenwrtNetmetrHandler(Handler, BaseOpenwrtHandler):
    uci = NetmetrUci()
    data = NetmetrDataFile()
    cmds = NetmetrCmds()

    @logger_wrapper(logger)
    def get_settings(self):
        return self.uci.get_settings()

    @logger_wrapper(logger)
    def update_settings(self, autostart_enabled, hours_to_run=None):
        return self.uci.update_settings(autostart_enabled, hours_to_run)

    @logger_wrapper(logger)
    def get_data(self):
        status, data = self.data.read_records()
        return {"status": status, "performed_tests": data}

    @logger_wrapper(logger)
    def download_data_trigger(self, exit_notify, reset_notify):
        return OpenwrtNetmetrHandler.cmds.download_data(exit_notify, reset_notify)

    @logger_wrapper(logger)
    def measure_and_download_data_trigger(self, notify, exit_notify, reset_notify):
        return OpenwrtNetmetrHandler.cmds.measure_and_download_data(
            notify, exit_notify, reset_notify)


foris_controller_modules/netmetr/handlers/mock.py
#
# foris-controller-netmetr-module
# Copyright (C) 2017 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

import logging
import random

from foris_controller.handler_base import BaseMockHandler
from foris_controller.utils import logger_wrapper

from .. import Handler

logger = logging.getLogger(__name__)


class MockNetmetrHandler(Handler, BaseMockHandler):
    autostart_enabled = True
    hours_to_run = list(range(0, 24, 2))
    sync_code = "1af827965d0d"

    @logger_wrapper(logger)
    def get_settings(self):
        return {
            "sync_code": self.sync_code,
            "autostart_enabled": self.autostart_enabled,
            "hours_to_run": self.hours_to_run,
        }

    @logger_wrapper(logger)
    def update_settings(self, autostart_enabled, hours_to_run=None):
        self.autostart_enabled = autostart_enabled
        if hours_to_run:
            self.hours_to_run = hours_to_run
        return True

    @logger_wrapper(logger)
    def get_data(self):
        return random.choice([
            {"status": "error", "performed_tests": []},
            {"status": "missing", "performed_tests": []},
            {"status": "ready", "performed_tests": [
                {
                    "speed_download": 23,
                    "speed_upload": 53,
                    "ping": 0.57,
                    "time": 1510669694,
                    "test_uuid": "563803cc-b43e-4b2e-bf9e-439b098fe35b"
                },
                {
                    "speed_download": 88,
                    "speed_upload": 66,
                    "ping": 0.81,
                    "time": 1510669699,
                    "test_uuid": "a975e00d-0bca-4871-8ce3-6e33c12311ef"
                }
            ]},
        ])

    @logger_wrapper(logger)
    def download_data_trigger(self, exit_notify, reset_notify):
        return "%032X" % random.randrange(2**32)

    @logger_wrapper(logger)
    def measure_and_download_data_trigger(self, notify, exit_notify, reset_notify):
        return "%032X" % random.randrange(2**32)
